### PR TITLE
feat: Support several counters for attempts

### PR DESCRIPTION
**Pull request by kletkavrubashku**
dev-testing is finished

### CODE REVIEW

File: vicodyn/src/vicodyn/balancer/simple.cpp [hunk 1]
```diff
@@ -14,16 +14,23 @@ simple_t::simple_t(context_t& ctx, peers_t& peers, asio::io_service& loop, const
...
     COCAINE_LOG_INFO(logger, "created simple balancer for app {}", app_name);
 }
 
 auto simple_t::choose_peer(const std::shared_ptr<request_context_t>& request_context,
                 const hpack::headers_t& /*headers*/, const std::string& /*event*/) -> std::shared_ptr<peer_t> {
+    static const std::string simple_counter_name = "simple_attempts";
+
+    auto& attempts = request_context->counter(simple_counter_name);
+    if (attempts >= attempt_limit) {
+        throw error_t("count of attempts is reached");
```
Comment (karitra) [inline vicodyn/src/vicodyn/balancer/simple.cpp]: Should be this exception reported to log? If yes, it is better to report additionally an`attempts_limit`.

File: vicodyn/include/cocaine/vicodyn/request_context.hpp [hunk 2]
```diff
@@ -38,11 +40,7 @@ class request_context_t: public std::enable_shared_from_this<request_context_t>
 
     auto peer_use_count(const std::string& peer_uuid) -> size_t;
 
-    auto register_retry() -> void;
-
-    auto retry_count() -> size_t;
-
-    auto custom_context() -> synchronized<boost::any>&;
+    auto counter(const std::string& name) -> std::atomic<std::size_t>&;
```
Comment (karitra) [inline vicodyn/include/cocaine/vicodyn/request_context.hpp]: It is not so important, but you are exposing internal realization details to class caller, after receiving the reference, caller would in general know that it is an atomic and could start treat it as specifically as atomic in caller code.

Comment (kletkavrubashku) [reply vicodyn/include/cocaine/vicodyn/request_context.hpp]: Yes, I'm agree. But usually a check of the counter and an increment of it are used very close, in the scope of one function. Current interface allows to use reference to a value and do the less count of getting counter by key. As a variant I can propose to return special class with method like "get" and method like "increment", but it seems like an overkill.

File: vicodyn/src/vicodyn/balancer/simple.cpp (hunk 1 above)
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/balancer/simple.cpp]: Ok

File: vicodyn/include/cocaine/vicodyn/request_context.hpp [hunk 3]
```diff
@@ -21,11 +21,13 @@ class request_context_t: public std::enable_shared_from_this<request_context_t>
     clock_t::time_point start_time_;
     std::atomic_flag closed_;
 
-    synchronized<std::vector<std::shared_ptr<peer_t>>> used_peers_;
-    synchronized<std::vector<checkpoint_t>> checkpoints_;
-    size_t retry_counter_;
+    using peer_array_t = std::vector<std::shared_ptr<peer_t>>;
+    using checkpoint_array_t = std::vector<checkpoint_t>;
+    using counter_map_t = std::map<std::string, std::atomic<std::size_t>>;
```
[the remark this answers is not in the document]
Comment (kletkavrubashku) [reply vicodyn/include/cocaine/vicodyn/request_context.hpp]: For regular ordering in log. Is it matter or not?

**Pull request merged**